src/main/java/com/expediagroup/hiveberg/IcebergInputFormat.java

```diff
@@ -94,22 +95,35 @@ public IcebergRecordReader(InputSplit split, JobConf conf) throws IOException {
     }
 
     private void initialise() {
-      //TODO Build different readers depending on file type
-      icebergSchema = table.schema();
-      avroSchema = getAvroSchema(AvroSchemaUtil.convert(icebergSchema, "temp").toString().replaceAll("-", "_"));
+      tasks = split.getTask().files().iterator();
+      nextTask();
+    }
 
-      CloseableIterable<GenericData.Record> reader = buildParquetReader(Files.localInput(split.getFile().path().toString()), icebergSchema, false);
+    private void nextTask(){
+      FileScanTask currentTask = tasks.next();
+      DataFile file = currentTask.file();
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), context);
+      Schema tableSchema = table.schema();
+      boolean reuseContainers = true; // FIXME: read from config
+
+      IcebergReaderFactory readerFactory = new IcebergReaderFactory();
+      reader = readerFactory.getReader(file, currentTask, inputFile, tableSchema, reuseContainers);
       recordIterator = reader.iterator();
     }
 
     @Override
-    public boolean next(Void key, AvroGenericRecordWritable value) {
+    public boolean next(Void key, IcebergWritable value) {
       if (recordIterator.hasNext()) {
-        GenericData.Record shadedRecord = recordIterator.next();
-        org.apache.avro.generic.GenericData.Record record = getUnshadedRecord(avroSchema, shadedRecord);
-        value.setRecord(record);
+        currentRecord = recordIterator.next();
+        value.setRecord(currentRecord);
         return true;
       }
+
+      if(tasks.hasNext()){
+        nextTask();
+        currentRecord = recordIterator.next();
+        value.setRecord(currentRecord);
```

Comment:
Shouldn't this return `true` now? I'm a little bit confused as to what's going on here and why we're looking in two places (`recordIterator` and `tasks`) for the next record...

Reply:
The way we've set it up so far is:
Each `IcebergSplit` contains one `CombinedScanTask` 
    -> Each `CombinedScanTask` can have multiple files to read (`FileScanTasks`). So the `tasks` iterator keeps track of which file we're currently looking at.
         -> But each file then also probably has multiple records, which is what the `recordIterator` is tracking. 

So when we're going through and fetching all the records, we need to check for all records in the first file, and then when we've finished with that one we check if we have any more files to check through until we've been through all the files (tasks)

Reply:
But you are also correct in that the second if should return true 🤣I don't catch that in my test case with my tiny table

Reply:
OK, I understand now. Doesn't this mean on line 124 we first need the same if check as we have on line 116? Or can we safely assume that a "new" task will always have some records? There's probably some refactoring that could be done here to make this a bit easier to understand but we can get it working first.

Reply:
I need to investigate a bit further into the Iceberg API but I think we can assume that a fresh task should have records to read, otherwise I don't think the file would get planned. But I'll try get a better understanding, and have a look at how to make this less confusing too